src/librustdoc/html/render/mod.rs

```diff
@@ -103,27 +102,42 @@ crate fn ensure_trailing_slash(v: &str) -> impl fmt::Display + '_ {
 crate struct Context<'tcx> {
     /// Current hierarchy of components leading down to what's currently being
     /// rendered
-    crate current: Vec<String>,
+    current: Vec<String>,
     /// The current destination folder of where HTML artifacts should be placed.
     /// This changes as the context descends into the module hierarchy.
-    crate dst: PathBuf,
+    dst: PathBuf,
     /// A flag, which when `true`, will render pages which redirect to the
     /// real location of an item. This is used to allow external links to
     /// publicly reused items to redirect to the right location.
-    crate render_redirect_pages: bool,
+    render_redirect_pages: bool,
     /// The map used to ensure all generated 'id=' attributes are unique.
-    id_map: Rc<RefCell<IdMap>>,
+    id_map: RefCell<IdMap>,
     /// Tracks section IDs for `Deref` targets so they match in both the main
     /// body and the sidebar.
-    deref_id_map: Rc<RefCell<FxHashMap<DefId, String>>>,
-    crate shared: Arc<SharedContext<'tcx>>,
-    all: Rc<RefCell<AllTypes>>,
-    /// Storage for the errors produced while generating documentation so they
-    /// can be printed together at the end.
-    crate errors: Rc<Receiver<String>>,
-    crate cache: Rc<Cache>,
+    deref_id_map: RefCell<FxHashMap<DefId, String>>,
+    /// Shared mutable state.
+    ///
+    /// Issue for improving the situation: [#82381][]
+    ///
+    /// [#82381]: https://github.com/rust-lang/rust/issues/82381
+    shared: Rc<SharedContext<'tcx>>,
+    /// The [`Cache`] used during rendering.
+    ///
+    /// Ideally the cache would be in [`SharedContext`], but it's mutated
+    /// between when the `SharedContext` is created and when `Context`
+    /// is created, so more refactoring would be needed.
+    ///
+    /// It's immutable once in `Context`, so it's not as bad that it's not in
+    /// `SharedContext`.
+    // FIXME: move `cache` to `SharedContext`
+    cache: Rc<Cache>,
 }
 
+// `Context` is cloned a lot, so we don't want the size to grow unexpectedly.
+#[cfg(target_arch = "x86_64")]
+rustc_data_structures::static_assert_size!(Context<'_>, 152);
```

Comment:
Oh, good point! Boxing should cut the size in half I think. Although it will still have to copy the buffer...

If that doesn't help, I have an idea for a nasty hack: add a custom `Clone` impl (or maybe make a new trait to avoid confusion) that clones everything except `id_map` and `deref_id_map` and instead creates empty hashmaps. I'm hoping to avoid that, but we'll see.

Reply:
If `id_map` and `deref_id_map` don't actually need to be part of Context, maybe you could take them out of the struct altogether and pass them as separate parameters? I think that would be a lot less confusing.

Reply:
I agree in theory, but `id_map` is populated before `Context` is created with data from `ExternalHtml` and I'm not sure right now how to refactor it to fix it. I think it would require threading `id_map` through a ton of places, and I'm hoping to prevent this PR from growing in size too much. I'll try making `deref_id_map` a parameter though.

Reply:
Actually, now that I think about it, if `id_map` is cleared before each item is rendered, maybe the stuff it's doing ahead-of-time is useless?

Reply:
Also, I'm thinking of renaming `Context` to `PageContext` to make its use clearer.

Reply:
You can make those changes in a parallel PR if you like, they don't have to necessarily be here.

Reply:
I had an idea for making this faster *and* cleaning up the use of `Clone` on the renderers 😁 

I'm trying that out now, we'll see how it goes.

Reply:
Tests passed on my local changes! 🎉

Reply:
Hmm, we don't seem to have any tests for `external_html`—I bet there is a pre-existing bug that stuff from `external_html` can cause HTML `id` conflicts.